crates/argmin/src/solver/itp/itp_method.rs

```diff
@@ -0,0 +1,354 @@
+
+    fn init(
+        &mut self,
+        problem: &mut Problem<O>,
+        // ItpRoot maintains its own state
+        state: IterState<F, (), (), (), (), F>,
+    ) -> Result<(IterState<F, (), (), (), (), F>, Option<KV>), Error> {
+        self.fa = problem.cost(&self.a)?;
+        self.fb = problem.cost(&self.b)?;
+        if self.fa * self.fb > float!(0.0) {
+            return Err(ItpRootError::WrongSign.into());
+        }
+        if self.tol < F::zero() {
+            return Err(ItpRootError::NegativeTol.into());
+        }
+        // This helps ensure the log evaluation is stable
+        if self.a > self.b {
+            return Err(ItpRootError::MinLargerThanMax.into());
+        }
+        // It's important to check this to verify n1o2 doesn't panic
+        if self.tol.is_zero() {
+            return Err(ItpRootError::ZeroTol.into());
+        }
+
+        // Preprocessing
+        self.n1o2 = ((self.b - self.a) / (float!(2.0) * self.tol)).log2();
+        self.nmax = self.n1o2 + self.n0;
+
+        Ok((state.param(self.b).cost(self.fb.abs()), None))
+    }
+
+    fn next_iter(
+        &mut self,
+        problem: &mut Problem<O>,
+        // ItpRoot maintains its own state
+        state: IterState<F, (), (), (), (), F>,
+    ) -> Result<(IterState<F, (), (), (), (), F>, Option<KV>), Error> {
+        // Note: the headers here match the steps outlined on Wikipedia, with variable names to
+        // match for clarity.
+
+        // Calculating Parameters
+        let b_minus_a = self.b - self.a;
+        let x1o2 = (self.a + self.b) * float!(0.5);
+        let r = self.tol * float!(2.0).powf(self.nmax + self.j) - b_minus_a * float!(0.5);
+        let delta = self.kappa1 * b_minus_a.powf(self.kappa2);
+
+        // Interpolation
+        let xf = (self.fb * self.a - self.fa * self.b) / (self.fb - self.fa);
+
+        // Truncation
+        let bisect_falsi_delta = x1o2 - xf;
+        let sigma = bisect_falsi_delta.signum();
+        let xt = if delta <= bisect_falsi_delta.abs() {
+            xf + sigma * delta
+        } else {
+            x1o2
+        };
+
+        // Projection
+        let xitp = if (xt - x1o2).abs() <= r {
+            xt
+        } else {
+            x1o2 - sigma * r
+        };
+
+        // Updating Interval
+        let fitp = problem.cost(&xitp)?;
+        if fitp > F::zero() {
+            self.b = xitp;
+            self.fb = fitp;
+        } else if fitp < F::zero() {
+            self.a = xitp;
+            self.fa = fitp;
+        } else {
+            self.a = xitp;
+            self.b = xitp;
+        }
+        self.j = self.j + float!(1.0);
+
+        // Solver loop termination
+        if (self.b - self.a) <= (float!(2.0) * self.tol) {
+            let sol = (self.a + self.b) * float!(0.5);
+            // TODO: This function evaluation serves no purpose other than to serve argmin's cost
+            // method on the state. It feels wasteful.
+            let f_sol = problem.cost(&sol)?;
```

Comment:
This bothers me, but maybe someone smarter than me might know a clever way to not do this.